src/frequenz/sdk/timeseries/_grid_frequency.py

```diff
@@ -0,0 +1,157 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Fetches the Grid Frequency."""
+
+
+import asyncio
+import logging
+
+from frequenz.channels import Receiver, Sender
+
+from ..actor import ChannelRegistry
+from ..microgrid import connection_manager
+from ..microgrid.component import Component, ComponentCategory, ComponentMetricId
+from ..timeseries._base_types import Sample
+from ..timeseries._quantities import Frequency
+
+
+def create_request(component_id: int) -> "ComponentMetricRequest":  # type: ignore
+    """Create a request for grid frequency.
+
+    Args:
+        component_id: The component id to use for the request.
+
+    Returns:
+        A component metric request for grid frequency.
+    """
+    # Imported here to avoid a circular import.
+    # pylint: disable=import-outside-toplevel
+    from ..actor import ComponentMetricRequest
+
+    return ComponentMetricRequest(
+        "grid-frequency", component_id, ComponentMetricId.FREQUENCY, None
+    )
+
+
+class GridFrequency:
+    """Grid Frequency."""
+
+    def __init__(
+        self,
+        data_sourcing_request_sender: Sender["ComponentMetricRequest"],  # type: ignore
+        channel_registry: ChannelRegistry,
+        component: Component,
+    ):
+        """Initialize the grid frequency formula generator.
+
+        Args:
+            channel_registry: The channel registry to use for the grid frequency.
+            data_sourcing_request_sender: The sender to use for requests.
+            component: The component to use for the grid frequency receiver. If not
+                provided, the first component that is either a meter, inverter or EV
+                charger will be used.
+        """
+        self._request_sender = data_sourcing_request_sender
+        self._channel_registry = channel_registry
+        self._component = component
+        self._component_metric_request = create_request(component.component_id)
+
+        self._task: None | asyncio.Task[None] = None
+
+    @property
+    def component(self) -> Component:
+        """The component that is used for grid frequency.
+
+        Returns:
+            The component that is used for grid frequency.
+        """
+        return self._component
+
+    def new_receiver(self) -> Receiver[Sample[Frequency]]:
+        """Create a receiver for grid frequency.
+
+        Returns:
+            A receiver that will receive grid frequency samples.
+        """
+        receiver = self._channel_registry.new_receiver(
+            self._component_metric_request.get_channel_name()
+        )
+
+        if not self._task:
+            self._task = asyncio.create_task(self._send_request())
+        else:
+            logging.info("Grid frequency request already sent: %s", self._component)
+
+        return receiver
```

Comment:
Just to put more accessible something that I said/asked somewhere else, this is hiding the `async` nature of creating a new receiver. Well, the receiver is created immediately, but it won't be receiving anything until the component metric request is fulfilled, so even if you have a receiver, some samples could be missed at first.

It was mentioned that for testing this is problematic because we need a sync point, because we want to verify we receive what we sent in tests. This is related to https://github.com/frequenz-floss/frequenz-channels-python/pull/154.

One solution would be to create a special receiver here that can be awaited to be ready, but I think that's a bit overkill for testing, same as adding the feature to channels. The other solution would be to add more mocking/instrumentation to tests, or to make tests resilient to the fact that a few first messages could be missed. To be honest, I think the later is the best option for tests, because this is the actual behavior of this method and we should test that.